### PR TITLE
feat(snapcraft): manage channel upload

### PR DESCRIPTION
**Pull request by guilhem**
By default, snap store manage "edge", "beta", "candidate" and "stable".
Default to all this channels when package is in grade stable. But only on
"edge" and "beta" for grade "devel".

This fix this TODO: https://github.com/goreleaser/goreleaser/blob/bf19dc1079887d31e58065771bb069953b1704d6/internal/pipe/snapcraft/snapcraft.go#L342

### CODE REVIEW

File: internal/pipe/snapcraft/snapcraft.go [hunk 1]
```diff
@@ -373,9 +384,31 @@ func link(src, dest string, mode os.FileMode) error {
...
+func copyFile(src, dst string, mode os.FileMode) error {
+	// Open original file
+	original, err := os.Open(src)
+	if err != nil {
+		return fmt.Errorf("fail to open '%s': %w", src, err)
+	}
+	defer original.Close()
+
+	// Create new file
+	new, err := os.OpenFile(dst, os.O_CREATE|os.O_WRONLY, mode)
```
Comment (caarlos0) [inline internal/pipe/snapcraft/snapcraft.go:403]: should probably truncate as well?

File: internal/pipe/snapcraft/snapcraft.go [hunk 2]
```diff
@@ -373,9 +384,31 @@ func link(src, dest string, mode os.FileMode) error {
...
-			return err
+		if err := copyFile(path, dst, mode); err != nil {
+			return fmt.Errorf("fail copy file '%s': %w", path, err)
 		}
-		return os.Chmod(dst, mode)
+		return nil
 	})
 }
+
+func copyFile(src, dst string, mode os.FileMode) error {
```
Comment (caarlos0) [inline internal/pipe/snapcraft/snapcraft.go:394]: curious why this change? did you find any issues with the hard linking one?

File: www/docs/customization/snapcraft.md [hunk 3]
```diff
@@ -60,12 +60,21 @@ snapcrafts:
     # store.
     description: This is the best drum roll application out there. Install it and awe!
 
+    # Channels in store where snap will be pushed.
+    # Default depends on grade.
```
Comment (caarlos0) [inline www/docs/customization/snapcraft.md:64]: probably worth saying here which channels are default for which grades?

**Comment by caarlos0** [comment]
Thanks for the PR! Just a couple of comments :)

File: internal/pipe/snapcraft/snapcraft.go (hunk 2 above)
Comment (guilhem) [reply internal/pipe/snapcraft/snapcraft.go:394]: Yes, hard link work only if both files are on the same FS.
My `/tmp` is managed outside, so tests are failing with:
```
failed to link completer: link testdata/foo-completer.bash /tmp/TestCompleter1448278326/001/dist/foo_arm6/prime/testdata/foo-completer.bash: invalid cross-device link
```
Generally, hardlinks are a bad idea.
I tried to use Symlink but `chmod` is inherited and failed.

File: www/docs/customization/snapcraft.md (hunk 3 above)
Comment (guilhem) [reply www/docs/customization/snapcraft.md:64]: Grade say something about it, but we may be more explicit indeed :)

File: internal/pipe/snapcraft/snapcraft.go (hunk 1 above)
Comment (guilhem) [reply internal/pipe/snapcraft/snapcraft.go:403]: nice shot ;)
Updating it

**Comment by guilhem** [comment]
@caarlos0 I updated to use a template for channels

**Review by caarlos0 — APPROVED** [review]
lgtm, thanks for the PR! 🤘
